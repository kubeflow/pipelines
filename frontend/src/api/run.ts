// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export class RunMetadata {
  public id = '';
  public name = '';
  public namespace = '';
  public created_at = '';
  public scheduled_at = '';
  public status = '';

  public static buildFromObject(run: any): RunMetadata {
    const newRunMetadata = new RunMetadata();
    newRunMetadata.id = run.id;
    newRunMetadata.name = run.name;
    newRunMetadata.namespace = run.namespace;
    newRunMetadata.created_at = run.created_at;
    newRunMetadata.scheduled_at = run.scheduled_at;
    newRunMetadata.status = run.status;
    return newRunMetadata;
  }
}

export class Run {
  public run = RunMetadata.buildFromObject({});
  public workflow = '';

  public static buildFromObject(oldRun: any): Run {
    const newRun = new Run();
    newRun.run = RunMetadata.buildFromObject(oldRun.run);
    newRun.workflow = oldRun.workflow;
    return newRun;
  }
}
