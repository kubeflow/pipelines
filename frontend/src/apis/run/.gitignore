wwwroot/*.js
node_modules
typings

# Ignore git_push script
git_push.sh
