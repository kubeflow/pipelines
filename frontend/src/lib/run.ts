export interface RunStep {
  name: string;
  start: number;
  end: number;
  state: string;
  outputs: string;
}

export interface Run {
  end: Date;
  id: number;
  instanceId: number;
  parameterValues: { [key: string]: number | string };
  progress: number;
  start: Date;
  state: 'not started' | 'running' | 'errored' | 'completed';
  steps: RunStep[];
}
