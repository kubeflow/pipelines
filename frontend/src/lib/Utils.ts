/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiRun } from '../apis/run';
import { ApiTrigger } from '../apis/job';
import { Workflow } from '../../third_party/argo-ui/argo_template';
import { isFunction } from 'lodash';
import { hasFinished, NodePhase } from './StatusUtils';

export const logger = {
  error: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.error(...args);
  },
  verbose: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.log(...args);
  },
};

export function formatDateString(date: Date | string | undefined): string {
  if (typeof date === 'string') {
    return new Date(date).toLocaleString();
  } else {
    return date ? date.toLocaleString() : '-';
  }
}

// TODO: add tests
export async function errorToMessage(error: any): Promise<string> {
  if (error instanceof Error) {
    return error.message;
  }

  if (error && error.text && isFunction(error.text)) {
    return await error.text();
  }

  return JSON.stringify(error) || '';
}

export function enabledDisplayString(trigger: ApiTrigger | undefined, enabled: boolean): string {
  if (trigger) {
    return enabled ? 'Yes' : 'No';
  }
  return '-';
}

function getDuration(start: Date, end: Date): string {
  let diff = end.getTime() - start.getTime();
  const sign = diff < 0 ? '-' : '';
  if (diff < 0) {
    diff *= -1;
  }
  const SECOND = 1000;
  const MINUTE = 60 * SECOND;
  const HOUR = 60 * MINUTE;
  const seconds = ('0' + Math.floor((diff / SECOND) % 60).toString()).slice(-2);
  const minutes = ('0' + Math.floor((diff / MINUTE) % 60).toString()).slice(-2);
  // Hours are the largest denomination, so we don't pad them
  const hours = Math.floor(diff / HOUR).toString();
  return `${sign}${hours}:${minutes}:${seconds}`;
}

export function getRunDuration(run?: ApiRun): string {
  if (!run || !run.created_at || !run.finished_at || !hasFinished(run.status as NodePhase)) {
    return '-';
  }

  // A bug in swagger-codegen causes the API to indicate that created_at and finished_at are Dates,
  // as they should be, when in reality they are transferred as strings.
  // See: https://github.com/swagger-api/swagger-codegen/issues/2776
  return getDuration(new Date(run.created_at), new Date(run.finished_at));
}

export function getRunDurationFromWorkflow(workflow?: Workflow): string {
  if (!workflow
    || !workflow.status
    || !workflow.status.startedAt
    || !workflow.status.finishedAt) {
    return '-';
  }

  return getDuration(new Date(workflow.status.startedAt), new Date(workflow.status.finishedAt));
}

export function s(items: any[] | number): string {
  const length = Array.isArray(items) ? items.length : items;
  return length === 1 ? '' : 's';
}
