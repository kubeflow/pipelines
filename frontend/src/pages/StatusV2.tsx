/*
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import ErrorIcon from '@material-ui/icons/Error';
import PendingIcon from '@material-ui/icons/Schedule';
import RunningIcon from 'src/icons/statusRunning';
import SkippedIcon from '@material-ui/icons/SkipNext';
import SuccessIcon from '@material-ui/icons/CheckCircle';
import CachedIcon from 'src/icons/statusCached';
import TerminatedIcon from 'src/icons/statusTerminated';
import Tooltip from '@material-ui/core/Tooltip';
import UnknownIcon from '@material-ui/icons/Help';
import { color } from 'src/Css';
import { logger, formatDateString } from 'src/lib/Utils';
import { checkIfTerminatedV2 } from 'src/lib/StatusUtils';
import { V2beta1RuntimeState } from 'src/apisv2beta1/run';
import { Execution } from 'src/third_party/mlmd/generated/ml_metadata/proto/metadata_store_pb';

export function statusToIcon(
  state?: V2beta1RuntimeState,
  startDate?: Date | string,
  endDate?: Date | string,
  nodeMessage?: string,
  mlmdState?: Execution.State,
): JSX.Element {
  state = checkIfTerminatedV2(state, nodeMessage);
  // tslint:disable-next-line:variable-name
  let IconComponent: any = UnknownIcon;
  let iconColor = color.inactive;
  let title = 'Unknown status';
  switch (state) {
    case V2beta1RuntimeState.FAILED:
      IconComponent = ErrorIcon;
      iconColor = color.errorText;
      title = 'Resource failed to execute';
      break;
    case V2beta1RuntimeState.PENDING:
      IconComponent = PendingIcon;
      iconColor = color.weak;
      title = 'Pending execution';
      break;
    case V2beta1RuntimeState.RUNNING:
      IconComponent = RunningIcon;
      iconColor = color.blue;
      title = 'Running';
      break;
    case V2beta1RuntimeState.CANCELING:
      IconComponent = RunningIcon;
      iconColor = color.blue;
      title = 'Run is canceling';
      break;
    case V2beta1RuntimeState.SKIPPED:
      IconComponent = SkippedIcon;
      title = 'Execution has been skipped for this resource';
      break;
    case V2beta1RuntimeState.SUCCEEDED:
      IconComponent = SuccessIcon;
      iconColor = color.success;
      title = 'Executed successfully';
      break;
    case V2beta1RuntimeState.CANCELED:
      IconComponent = TerminatedIcon;
      iconColor = color.terminated;
      title = 'Run was manually canceled';
      break;
    case V2beta1RuntimeState.RUNTIMESTATEUNSPECIFIED:
      break;
    default:
      logger.verbose('Unknown state:', state);
  }
  // TODO(jlyaoyuli): Additional changes is probably needed after Status IR integration.
  if (mlmdState === Execution.State.CACHED) {
    IconComponent = CachedIcon;
    iconColor = color.success;
    title = 'Execution was skipped and outputs were taken from cache';
  }
  return (
    <Tooltip
      title={
        <div>
          <div>{title}</div>
          {/* These dates may actually be strings, not a Dates due to a bug in swagger's handling of dates */}
          {startDate && <div>Start: {formatDateString(startDate)}</div>}
          {endDate && <div>End: {formatDateString(endDate)}</div>}
        </div>
      }
    >
      <div>
        <IconComponent
          data-testid='node-status-sign'
          style={{ color: iconColor, height: 18, width: 18 }}
        />
      </div>
    </Tooltip>
  );
}
