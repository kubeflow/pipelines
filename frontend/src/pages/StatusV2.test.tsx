/*
 * Copyright 2023 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as Utils from 'src/lib/Utils';
import { statusToIcon } from './StatusV2';
import { shallow } from 'enzyme';
import { V2beta1RuntimeState } from 'src/apisv2beta1/run';

describe('Status', () => {
  // We mock this because it uses toLocaleDateString, which causes mismatches between local and CI
  // test enviroments
  const formatDateStringSpy = jest.spyOn(Utils, 'formatDateString');

  const startDate = new Date('Wed Jan 2 2019 9:10:11 GMT-0800');
  const endDate = new Date('Thu Jan 3 2019 10:11:12 GMT-0800');

  beforeEach(() => {
    formatDateStringSpy.mockImplementation((date: Date) => {
      return date === startDate ? '1/2/2019, 9:10:11 AM' : '1/3/2019, 10:11:12 AM';
    });
  });

  describe('statusToIcon', () => {
    it('handles an unknown state', () => {
      const consoleSpy = jest.spyOn(console, 'log').mockImplementationOnce(() => null);
      const tree = shallow(statusToIcon('bad state' as any));
      expect(tree).toMatchSnapshot();
      expect(consoleSpy).toHaveBeenLastCalledWith('Unknown state:', 'bad state');
    });

    it('handles an undefined state', () => {
      const consoleSpy = jest.spyOn(console, 'log').mockImplementationOnce(() => null);
      const tree = shallow(statusToIcon(/* no phase */));
      expect(tree).toMatchSnapshot();
      expect(consoleSpy).toHaveBeenLastCalledWith('Unknown state:', undefined);
    });

    // TODO: Enable this test after react-scripts is upgraded to v4.0.0
    // it('react testing for ERROR state', async () => {
    //   const { findByText, getByTestId } = render(
    //     statusToIcon(NodePhase.ERROR),
    //   );

    //   fireEvent.mouseOver(getByTestId('node-status-sign'));
    //   findByText('Error while running this resource');
    // });

    it('handles FAILED state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.FAILED));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <pure(ErrorIcon)
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#d50000",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles PENDING state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.PENDING));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <pure(ScheduleIcon)
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#9aa0a6",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles RUNNING state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.RUNNING));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <StatusRunning
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#4285f4",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles CANCELING state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.CANCELING));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <StatusRunning
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#4285f4",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles SKIPPED state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SKIPPED));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <pure(SkipNextIcon)
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#5f6368",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles SUCCEEDED state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SUCCEEDED));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <pure(CheckCircleIcon)
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#34a853",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('handles CANCELED state', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.CANCELED));
      expect(tree.find('div')).toMatchInlineSnapshot(`
        <div>
          <StatusRunning
            data-testid="node-status-sign"
            style={
              Object {
                "color": "#80868b",
                "height": 18,
                "width": 18,
              }
            }
          />
        </div>
      `);
    });

    it('displays start and end dates if both are provided', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SUCCEEDED, startDate, endDate));
      expect(tree).toMatchSnapshot();
    });

    it('does not display a end date if none was provided', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SUCCEEDED, startDate));
      expect(tree).toMatchSnapshot();
    });

    it('does not display a start date if none was provided', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SUCCEEDED, undefined, endDate));
      expect(tree).toMatchSnapshot();
    });

    it('does not display any dates if neither was provided', () => {
      const tree = shallow(statusToIcon(V2beta1RuntimeState.SUCCEEDED /* No dates */));
      expect(tree).toMatchSnapshot();
    });

    Object.keys(V2beta1RuntimeState).map(status =>
      it('renders an icon with tooltip for phase: ' + status, () => {
        const tree = shallow(statusToIcon(V2beta1RuntimeState[status]));
        expect(tree).toMatchSnapshot();
      }),
    );
  });
});
