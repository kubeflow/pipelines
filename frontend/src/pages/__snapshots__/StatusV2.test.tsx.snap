// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status statusToIcon displays start and end dates if both are provided 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(52, 168, 83); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon does not display a end date if none was provided 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(52, 168, 83); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon does not display a start date if none was provided 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(52, 168, 83); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon does not display any dates if neither was provided 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(52, 168, 83); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon handles CANCELED state 1`] = `
<div
  class=""
  data-mui-internal-clone-element="true"
>
  <svg
    height="18"
    viewBox="0 0 18 18"
    width="18"
  >
    <g
      fill="none"
      fill-rule="evenodd"
      stroke="none"
      stroke-width="1"
    >
      <g
        transform="translate(-1.000000, -1.000000)"
      >
        <polygon
          points="0 0 18 0 18 18 0 18"
        />
        <path
          d="M8.9925,1.5 C4.8525,1.5 1.5,4.86 1.5,9 C1.5,13.14 4.8525,16.5 8.9925,16.5 C13.14,16.5 16.5,13.14 16.5,9 C16.5,4.86 13.14,1.5 8.9925,1.5 Z M9,15 C5.685,15 3,12.315 3,9 C3,5.685 5.685,3 9,3 C12.315,3 15,5.685 15,9 C15,12.315 12.315,15 9,15 Z"
          fill="#80868b"
          fill-rule="nonzero"
        />
        <polygon
          fill="#80868b"
          fill-rule="nonzero"
          points="6 6 12 6 12 12 6 12"
        />
      </g>
    </g>
  </svg>
</div>
`;

exports[`Status statusToIcon handles CANCELING state 1`] = `
<div
  class=""
  data-mui-internal-clone-element="true"
>
  <svg
    height="18"
    viewBox="0 0 18 18"
    width="18"
  >
    <g
      fill="#4285f4"
      fill-rule="nonzero"
      transform="translate(-450, -307)"
    >
      <g
        transform="translate(450, 266)"
      >
        <g
          transform="translate(0, 41)"
        >
          <path
            d="M9,4 C6.23857143,4 4,6.23857143 4,9 C4,11.7614286 6.23857143,14 9,14 C11.7614286,14 14,11.7614286 14,9 C14,8.40214643 13.8950716,7.8288007 13.702626,7.29737398 L15.2180703,5.78192967 C15.7177126,6.74539838 16,7.83973264 16,9 C16,12.866 12.866,16 9,16 C5.134,16 2,12.866 2,9 C2,5.134 5.134,2 9,2 C10.933,2 12.683,2.7835 13.94975,4.05025 L12.7677679,5.23223214 L9,9 L9,4 Z"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Status statusToIcon handles RUNNING state 1`] = `
<div
  class=""
  data-mui-internal-clone-element="true"
>
  <svg
    height="18"
    viewBox="0 0 18 18"
    width="18"
  >
    <g
      fill="#4285f4"
      fill-rule="nonzero"
      transform="translate(-450, -307)"
    >
      <g
        transform="translate(450, 266)"
      >
        <g
          transform="translate(0, 41)"
        >
          <path
            d="M9,4 C6.23857143,4 4,6.23857143 4,9 C4,11.7614286 6.23857143,14 9,14 C11.7614286,14 14,11.7614286 14,9 C14,8.40214643 13.8950716,7.8288007 13.702626,7.29737398 L15.2180703,5.78192967 C15.7177126,6.74539838 16,7.83973264 16,9 C16,12.866 12.866,16 9,16 C5.134,16 2,12.866 2,9 C2,5.134 5.134,2 9,2 C10.933,2 12.683,2.7835 13.94975,4.05025 L12.7677679,5.23223214 L9,9 L9,4 Z"
          />
        </g>
      </g>
    </g>
  </svg>
</div>
`;

exports[`Status statusToIcon handles an undefined state 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(95, 99, 104); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon handles an unknown state 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(95, 99, 104); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: CANCELED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      height="18"
      viewBox="0 0 18 18"
      width="18"
    >
      <g
        fill="none"
        fill-rule="evenodd"
        stroke="none"
        stroke-width="1"
      >
        <g
          transform="translate(-1.000000, -1.000000)"
        >
          <polygon
            points="0 0 18 0 18 18 0 18"
          />
          <path
            d="M8.9925,1.5 C4.8525,1.5 1.5,4.86 1.5,9 C1.5,13.14 4.8525,16.5 8.9925,16.5 C13.14,16.5 16.5,13.14 16.5,9 C16.5,4.86 13.14,1.5 8.9925,1.5 Z M9,15 C5.685,15 3,12.315 3,9 C3,5.685 5.685,3 9,3 C12.315,3 15,5.685 15,9 C15,12.315 12.315,15 9,15 Z"
            fill="#80868b"
            fill-rule="nonzero"
          />
          <polygon
            fill="#80868b"
            fill-rule="nonzero"
            points="6 6 12 6 12 12 6 12"
          />
        </g>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: CANCELING 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      height="18"
      viewBox="0 0 18 18"
      width="18"
    >
      <g
        fill="#4285f4"
        fill-rule="nonzero"
        transform="translate(-450, -307)"
      >
        <g
          transform="translate(450, 266)"
        >
          <g
            transform="translate(0, 41)"
          >
            <path
              d="M9,4 C6.23857143,4 4,6.23857143 4,9 C4,11.7614286 6.23857143,14 9,14 C11.7614286,14 14,11.7614286 14,9 C14,8.40214643 13.8950716,7.8288007 13.702626,7.29737398 L15.2180703,5.78192967 C15.7177126,6.74539838 16,7.83973264 16,9 C16,12.866 12.866,16 9,16 C5.134,16 2,12.866 2,9 C2,5.134 5.134,2 9,2 C10.933,2 12.683,2.7835 13.94975,4.05025 L12.7677679,5.23223214 L9,9 L9,4 Z"
            />
          </g>
        </g>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: FAILED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(213, 0, 0); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-2h2zm0-4h-2V7h2z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: PAUSED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(95, 99, 104); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: PENDING 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(154, 160, 166); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2M12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8"
      />
      <path
        d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: RUNNING 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      height="18"
      viewBox="0 0 18 18"
      width="18"
    >
      <g
        fill="#4285f4"
        fill-rule="nonzero"
        transform="translate(-450, -307)"
      >
        <g
          transform="translate(450, 266)"
        >
          <g
            transform="translate(0, 41)"
          >
            <path
              d="M9,4 C6.23857143,4 4,6.23857143 4,9 C4,11.7614286 6.23857143,14 9,14 C11.7614286,14 14,11.7614286 14,9 C14,8.40214643 13.8950716,7.8288007 13.702626,7.29737398 L15.2180703,5.78192967 C15.7177126,6.74539838 16,7.83973264 16,9 C16,12.866 12.866,16 9,16 C5.134,16 2,12.866 2,9 C2,5.134 5.134,2 9,2 C10.933,2 12.683,2.7835 13.94975,4.05025 L12.7677679,5.23223214 L9,9 L9,4 Z"
            />
          </g>
        </g>
      </g>
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: RUNTIMESTATEUNSPECIFIED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(95, 99, 104); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: SKIPPED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(95, 99, 104); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="m6 18 8.5-6L6 6zM16 6v12h2V6z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Status statusToIcon renders an icon with tooltip for phase: SUCCEEDED 1`] = `
<DocumentFragment>
  <div
    class=""
    data-mui-internal-clone-element="true"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="node-status-sign"
      focusable="false"
      style="color: rgb(52, 168, 83); height: 18px; width: 18px;"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
      />
    </svg>
  </div>
</DocumentFragment>
`;
